simulation/halsim_ws_core/src/main/native/cpp/WSProvider_Analog.cpp

```diff
@@ -0,0 +1,154 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2017-2020 FIRST. All Rights Reserved.                        */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#include "WSProvider_Analog.h"
+
+#include <hal/Ports.h>
+#include <hal/simulation/AnalogInData.h>
+#include <hal/simulation/AnalogOutData.h>
+
+namespace wpilibws {
+
+void HALSimWSProviderAnalogIn::Initialize(WSRegisterFunc webRegisterFunc) {
+  CreateProviders<HALSimWSProviderAnalogIn>("AI", HAL_GetNumAnalogInputs(),
+                                            webRegisterFunc);
+}
+
+void HALSimWSProviderAnalogIn::RegisterCallbacks() {
+  m_initCbKey = HALSIM_RegisterAnalogInInitializedCallback(
+      m_channel,
+      [](const char* name, void* param, const struct HAL_Value* value) {
+        static_cast<HALSimWSProviderAnalogIn*>(param)->ProcessHalCallback(
+            {{"<init", static_cast<bool>(value->data.v_boolean)}});
+      },
+      this, true);
```

Comment:
Just a suggestion: it might be beneficial to macro-ize this and similar files to reduce the amount of copy and paste.  E.g. then this becomes:
```
#define REGISTER(halsim, jsonid, type) \
  HALSIM_Register##halsim##Callback( \
      m_channel, \
      [](const char* name, void* param, const struct HAL_Value* value) { \
        static_cast<HALSimWSProviderAnalogIn*>(param)->ProcessHalCallback( \
            {{jsonid, static_cast<bool>(value->data.v_##type)}}); \
      }, \
      this, true);

m_initCbKey = REGISTER(AnalogInInitialized, "<init", boolean);
m_avgbitsCbKey = REGISTER(AnalogInAverageBits, "<avg_bits", int);
m_oversampleCbKey = REGISTER(AnalogInOversampleBits, "<oversample_bits", int);

etc.
```

Reply:
Ooh good point. I’ll try that out in the next commit. Also incoming shortly, a very rudimentary   SendFile replacement that also works on windows